docs/zq/language.md

```diff
@@ -1,341 +1,2091 @@
 
-## Regular Expressions
+## 1. Introduction
 
-TBD: this is linked from below
-
-## Globs
+The Zed language is a query language for search, analytics,
+and transformation inspired by the
+[pipeline pattern](https://en.wikipedia.org/wiki/Tacit_programming)
+of the traditional Unix shell.
+Like a Unix pipeline, a query is expressed as a data source followed
+by a number of commands:
+```
+command | command | command | ...
+```
+However, in Zed, the entities that transform data are called
+"operators" instead of "commands" and unlike Unix pipelines,
+the streams of data in a Zed query
+are typed data sequences that adhere to the
+[Zed data model](../formats/zed.md#the-zed-data-model-specification).
+Moreover, Zed sequences can be forked and joined:
+```
+operator
+| operator
+| fork (
+  => operator | ...
+  => operator | ...
+)
+| join | ...
+```
+Here, Zed programs can include multiple data sources and splitting operations
+where multiple paths run in parallel and paths can be combined (in an
+undefined order), merged (in a defined order) by one or more sort keys,
+or joined using relational-style join logic.
 
-TBD: this is linked from below
+Generally speaking, a [flow graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph)
+defines a directed acyclic graph (DAG) composed
+of data sources and operator nodes.  The Zed syntax leverages "fat arrows",
+i.e., `=>`, to indicate the start of a parallel legs of the data flow.
 
-## Record Mutations
+That said, the Zed language is
+[declarative](https://en.wikipedia.org/wiki/Declarative_programming)
+and the Zed compiler optimizes the data flow computation
+&mdash; e.g., often implementing a Zed program differently than
+the flow implied by the pipeline yet reaching the same result &mdash;
+much as a modern SQL engine optimizes a declarative SQL query.
 
-TBD: this is linked from below
+Zed is also intended to provide a seamless transition from a simple search experience
+(e.g., typed into a search bar or as the query argument of the  `zq` command-line
+tool) to more a complex analytics experience composed of complex joins and aggregations
+where the Zed language source text would typically be authored in a editor and
+managed under source-code control.
 
-## Type Fusion
+Like an email or Web search, a simple keyword search is just the word itself,
+e.g.,
+```
+sample.com
+```
+is a search for the string "sample.com" and
+```
+sample.com urgent
+```
+is a search for values with the both strings "sample.com" and "urgent" present.
 
-TBD: this is linked from below
+Unlike typical log search systems, the Zed language operators are uniform:
+you can specify an operator including keyword search terms, Boolean predicates,
+etc. using the same syntax at any point in the pipeline as
+[described below](#7-search-expressions)
 
-## Record Literal
+For example,
+the predicate `message_length > 100` can simply be tacked onto the keyword search
+from above, e.g.,
+```
+sample.com urgent message_length > 100
```

Comment:
```suggestion
example.com urgent message_length > 100
```
or,
```suggestion
urgent sample.com message_length > 100
```
or,
```suggestion
'sample.com' urgent message_length > 100
```
(`sample` is reserved for the zed operator - renaming to `example.com` might make most sense?)